sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java

```diff
@@ -1703,4 +1774,498 @@ public void onCompletion(RecordMetadata metadata, Exception exception) {
+    //     Exactly-once semantics are achieved by writing records in the strict order of
+    //     these checkpointed sequence ids.
+    //
+    // Parallelism for B and C is fixed to 'numShards', which defaults to number of partitions
+    // for the topic. A few reasons for that:
+    //  - B & C implement their functionality using per-key state. Shard id makes it independent
+    //    of cardinality of user key.
+    //  - We create one producer per shard, and its 'transactional id' is based on shard id. This
+    //    requires that number of shards to be finite. This also helps with batching. and avoids
+    //    initializing producers and transactions.
+    //  - Most importantly, each of sharded writers stores 'next message id' in partition
+    //    metadata, which is committed atomically with Kafka transactions. This is critical
+    //    to handle retries of C correctly. Initial testing showed number of shards could be
+    //    larger than number of partitions for the topic.
+    //
+    // Number of shards can change across multiple runs of a pipeline (job upgrade in Dataflow).
+    //
+
+    private final Write<K, V> spec;
+
+    EOSWrite(Write<K, V> spec) {
+      this.spec = spec;
+    }
+
+    @Override
+    public PCollection<Void> expand(PCollection<KV<K, V>> input) {
+
+      int numShards = spec.getNumShards();
+      if (numShards <= 0) {
+        try (Consumer<?, ?> consumer = openConsumer(spec)) {
+          numShards = consumer.partitionsFor(spec.getTopic()).size();
+          LOG.info("Using {} shards for exactly-once, matching number of partitions for topic '{}'",
+                   numShards, spec.getTopic());
+        }
+      }
+      checkState(numShards > 0, "Could not set number of shards");
+
+      return input
+          .apply(Window.<KV<K, V>>into(new GlobalWindows()) // Everything into global window.
+                     .triggering(Repeatedly.forever(AfterPane.elementCountAtLeast(1)))
+                     .discardingFiredPanes())
+          .apply(String.format("Shuffle across %d shards", numShards),
+                 ParDo.of(new EOSReshard<K, V>(numShards)))
+          .apply("Persist sharding", GroupByKey.<Integer, KV<K, V>>create())
+          .apply("Assign sequential ids", ParDo.of(new EOSSequencer<K, V>()))
+          .apply("Persist ids", GroupByKey.<Integer, KV<Long, KV<K, V>>>create())
+          .apply(String.format("Write to Kafka topic '%s'", spec.getTopic()),
+                 ParDo.of(new KafkaEOWriter<>(spec, input.getCoder())));
+    }
+  }
+
+  /**
+   * Shuffle messages assigning each randomly to a shard.
+   */
+  private static class EOSReshard<K, V> extends DoFn<KV<K, V>, KV<Integer, KV<K, V>>> {
+    private final int numShards;
+
+    EOSReshard(int numShards) {
+      this.numShards = numShards;
+    }
+
+    @ProcessElement
+    public void processElement(ProcessContext ctx) {
+      int shard = ThreadLocalRandom.current().nextInt(numShards);
```

Comment:
<!--new_thread; commit:27f622c4e2c0fabf68d43da4e4aa381f71f82365; resolved:0-->
call ThreadLocalRandom in @Setup, and here just use (shard ++) % numShards.